File: src/calculations/units.ts
import type {
  VolumeType,
  MassType,
  TimeType,
  TemperatureType,
  ColorType,
  PressureType,
  GravityType,
  ConcentrationType,
  SpecificVolumeType,
  AcidityType,
  CarbonationType,
  BitternessType,
  PercentType,
} from "@beerjson/beerjson";

/**
 * Unit conversion utilities for BeerJSON types.
 * All conversions support the complete set of units defined in the BeerJSON format.
 */

// ============================================================================
// VOLUME CONVERSIONS
// ============================================================================

/**
 * Convert any VolumeType to milliliters (ml)
 */
export function volumeToMilliliters(volume: VolumeType): number {
  switch (volume.unit) {
    case "ml":
      return volume.value;
    case "l":
      return volume.value * 1000;
    case "tsp":
      return volume.value * 4.92892;
    case "tbsp":
      return volume.value * 14.7868;
    case "floz":
      return volume.value * 29.5735;
    case "cup":
      return volume.value * 236.588;
    case "pt":
      return volume.value * 473.176;
    case "qt":
      return volume.value * 946.353;
    case "gal":
      return volume.value * 3785.41;
    case "bbl":
      return volume.value * 117347.77;
    case "ifloz":
      return volume.value * 28.4131;
    case "ipt":
      return volume.value * 568.261;
    case "iqt":
      return volume.value * 1136.52;
    case "igal":
      return volume.value * 4546.09;
    case "ibbl":
      return volume.value * 163659.24;
    default:
      throw Error(`Unrecognized volume unit: ${volume.unit}`);
  }
}

/**
 * Convert any VolumeType to liters (l)
 */
export function volumeToLiters(volume: VolumeType): number {
  return volumeToMilliliters(volume) / 1000;
}

/**
 * Convert any VolumeType to US gallons (gal)
 */
export function volumeToGallons(volume: VolumeType): number {
  return volumeToMilliliters(volume) / 3785.41;
}

/**
 * Convert any VolumeType to Imperial gallons (igal)
 */
export function volumeToImperialGallons(volume: VolumeType): number {
  return volumeToMilliliters(volume) / 4546.09;
}

// ============================================================================
// MASS CONVERSIONS
// ============================================================================

/**
 * Convert any MassType to grams (g)
 */
export function massToGrams(mass: MassType): number {
  switch (mass.unit) {
    case "mg":
      return mass.value / 1000;
    case "g":
      return mass.value;
    case "kg":
      return mass.value * 1000;
    case "lb":
      return mass.value * 453.592;
    case "oz":
      return mass.value * 28.3495;
    default:
      throw Error(`Unrecognized mass unit: ${mass.unit}`);
  }
}

/**
 * Convert any MassType to kilograms (kg)
 */
export function massToKilograms(mass: MassType): number {
  return massToGrams(mass) / 1000;
}

/**
 * Convert any MassType to pounds (lb)
 */
export function massToPounds(mass: MassType): number {
  return massToGrams(mass) / 453.592;
}

/**
 * Convert any MassType to ounces (oz)
 */
export function massToOunces(mass: MassType): number {
  return massToGrams(mass) / 28.3495;
}

// ============================================================================
// TIME CONVERSIONS
// ============================================================================

/**
 * Convert any TimeType to seconds (sec)
 */
export function timeToSeconds(time: TimeType): number {
  switch (time.unit) {
    case "sec":
      return time.value;
    case "min":
      return time.value * 60;
    case "hr":
      return time.value * 3600;
    case "day":
      return time.value * 86400;
    case "week":
      return time.value * 604800;
    default:
      throw Error(`Unrecognized time unit: ${time.unit}`);
  }
}

/**
 * Convert any TimeType to minutes (min)
 */
export function timeToMinutes(time: TimeType): number {
  return timeToSeconds(time) / 60;
}

/**
 * Convert any TimeType to hours (hr)
 */
export function timeToHours(time: TimeType): number {
  return timeToSeconds(time) / 3600;
}

/**
 * Convert any TimeType to days (day)
 */
export function timeToDays(time: TimeType): number {
  return timeToSeconds(time) / 86400;
}

// ============================================================================
// TEMPERATURE CONVERSIONS
// ============================================================================

/**
 * Convert any TemperatureType to Celsius (C)
 */
export function temperatureToCelsius(temp: TemperatureType): number {
  switch (temp.unit) {
    case "C":
      return temp.value;
    case "F":
      return (temp.value - 32) * (5 / 9);
    default:
      throw Error(`Unrecognized temperature unit: ${temp.unit}`);
  }
}

/**
 * Convert any TemperatureType to Fahrenheit (F)
 */
export function temperatureToFahrenheit(temp: TemperatureType): number {
  switch (temp.unit) {
    case "C":
      return temp.value * (9 / 5) + 32;
    case "F":
      return temp.value;
    default:
      throw Error(`Unrecognized temperature unit: ${temp.unit}`);
  }
}

// ============================================================================
// COLOR CONVERSIONS
// ============================================================================

/**
 * Convert SRM to EBC
 */
export function srmToEbc(srm: number): number {
  return 1.97 * srm;
}

/**
 * Convert EBC to SRM
 */
export function ebcToSrm(ebc: number): number {
  return ebc / 1.97;
}

/**
 * Convert Lovibond to SRM
 */
export function lovibondToSrm(lovibond: number): number {
  return 1.3546 * lovibond - 0.76;
}

/**
 * Convert SRM to Lovibond
 */
export function srmToLovibond(srm: number): number {
  return (srm + 0.76) / 1.3546;
}

/**
 * Convert any ColorType to SRM
 */
export function colorToSrm(color: ColorType): number {
  switch (color.unit) {
    case "SRM":
      return color.value;
    case "EBC":
      return ebcToSrm(color.value);
    case "Lovi":
      return lovibondToSrm(color.value);
    default:
      throw Error(`Unrecognized color unit: ${color.unit}`);
  }
}

/**
 * Convert any ColorType to EBC
 */
export function colorToEbc(color: ColorType): number {
  return srmToEbc(colorToSrm(color));
}

/**
 * Convert any ColorType to Lovibond
 */
export function colorToLovibond(color: ColorType): number {
  return srmToLovibond(colorToSrm(color));
}

// ============================================================================
// PRESSURE CONVERSIONS
// ============================================================================

/**
 * Convert any PressureType to kilopascals (kPa)
 */
export function pressureToKilopascals(pressure: PressureType): number {
  switch (pressure.unit) {
    case "kPa":
      return pressure.value;
    case "psi":
      return pressure.value * 6.89476;
    case "bar":
      return pressure.value * 100;
    default:
      throw Error(`Unrecognized pressure unit: ${pressure.unit}`);
  }
}

/**
 * Convert any PressureType to PSI
 */
export function pressureToPsi(pressure: PressureType): number {
  return pressureToKilopascals(pressure) / 6.89476;
}

/**
 * Convert any PressureType to bar
 */
export function pressureToBar(pressure: PressureType): number {
  return pressureToKilopascals(pressure) / 100;
}

// ============================================================================
// GRAVITY CONVERSIONS
// ============================================================================

/**
 * Convert any GravityType to specific gravity (sg)
 */
export function gravityToSG(gravity: GravityType): number {
  switch (gravity.unit) {
    case "sg":
      return gravity.value;
    case "plato":
      // Plato to SG: SG = 1 + (Plato / (258.6 - ((Plato / 258.2) * 227.1)))
      return (
        1 +
        gravity.value / (258.6 - (gravity.value / 258.2) * 227.1)
      );
    case "brix":
      // Brix to SG (using same formula as Plato, as they're very similar)
      return (
        1 +
        gravity.value / (258.6 - (gravity.value / 258.2) * 227.1)
      );
    default:
      throw Error(`Unrecognized gravity unit: ${gravity.unit}`);
  }
}

/**
 * Convert any GravityType to Plato
 */
export function gravityToPlato(gravity: GravityType): number {
  switch (gravity.unit) {
    case "sg":
      // SG to Plato: Plato = -616.868 + 1111.14 * SG - 630.272 * SG^2 + 135.997 * SG^3
      const sg = gravity.value;
      return (
        -616.868 +
        1111.14 * sg -
        630.272 * Math.pow(sg, 2) +
        135.997 * Math.pow(sg, 3)
      );
    case "plato":
      return gravity.value;
    case "brix":
      return gravity.value;
    default:
      throw Error(`Unrecognized gravity unit: ${gravity.unit}`);
  }
}

/**
 * Convert any GravityType to Brix
 */
export function gravityToBrix(gravity: GravityType): number {
  return gravityToPlato(gravity);
}

// ============================================================================
// CONCENTRATION CONVERSIONS
// ============================================================================

/**
 * Convert any ConcentrationType to mg/l (milligrams per liter)
 */
export function concentrationToMgPerL(concentration: ConcentrationType): number {
  switch (concentration.unit) {
    case "mg/l":
      return concentration.value;
    case "ppm":
      return concentration.value; // ppm is equivalent to mg/l for aqueous solutions
    case "ppb":
      return concentration.value / 1000;
    default:
      throw Error(`Unrecognized concentration unit: ${concentration.unit}`);
  }
}

/**
 * Convert any ConcentrationType to ppm (parts per million)
 */
export function concentrationToPpm(concentration: ConcentrationType): number {
  return concentrationToMgPerL(concentration);
}

/**
 * Convert any ConcentrationType to ppb (parts per billion)
 */
export function concentrationToPpb(concentration: ConcentrationType): number {
  return concentrationToMgPerL(concentration) * 1000;
}

// ============================================================================
// SPECIFIC VOLUME CONVERSIONS
// ============================================================================

/**
 * Convert any SpecificVolumeType to l/kg (liters per kilogram)
 */
export function specificVolumeToLPerKg(
  specificVolume: SpecificVolumeType
): number {
  switch (specificVolume.unit) {
    case "l/kg":
      return specificVolume.value;
    case "l/g":
      return specificVolume.value * 1000;
    case "qt/lb":
      return specificVolume.value * 2.08635;
    case "gal/lb":
      return specificVolume.value * 8.34540;
    case "gal/oz":
      return specificVolume.value * 133.526;
    case "floz/oz":
      return specificVolume.value * 1.04318;
    case "m^3/kg":
      return specificVolume.value * 1000;
    case "ft^3/lb":
      return specificVolume.value * 62.4279;
    default:
      throw Error(`Unrecognized specific volume unit: ${specificVolume.unit}`);
  }
}

/**
 * Convert any SpecificVolumeType to qt/lb (quarts per pound)
 */
export function specificVolumeToQtPerLb(
  specificVolume: SpecificVolumeType
): number {
  return specificVolumeToLPerKg(specificVolume) / 2.08635;
}

/**
 * Convert any SpecificVolumeType to gal/lb (gallons per pound)
 */
export function specificVolumeToGalPerLb(
  specificVolume: SpecificVolumeType
): number {
  return specificVolumeToLPerKg(specificVolume) / 8.34540;
}

/**
 * Convert any SpecificVolumeType to gal/kg (gallons per kilogram)
 */
export function specificVolumeToGallonsPerKilogram(
  specificVolume: SpecificVolumeType
): number {
  return specificVolumeToLPerKg(specificVolume) * 0.264172;
}

// ============================================================================
// ACIDITY CONVERSIONS
// ============================================================================

/**
 * Convert any AcidityType to pH value
 */
export function acidityToPh(acidity: AcidityType): number {
  switch (acidity.unit) {
    case "pH":
      return acidity.value;
    default:
      throw Error(`Unrecognized acidity unit: ${acidity.unit}`);
  }
}

// ============================================================================
// CARBONATION CONVERSIONS
// ============================================================================

/**
 * Convert any CarbonationType to volumes of CO2
 */
export function carbonationToVolumes(carbonation: CarbonationType): number {
  switch (carbonation.unit) {
    case "vols":
      return carbonation.value;
    case "g/l":
      // g/l to volumes: 1 volume = 1.96 g/l at STP
      return carbonation.value / 1.96;
    default:
      throw Error(`Unrecognized carbonation unit: ${carbonation.unit}`);
  }
}

/**
 * Convert any CarbonationType to g/l (grams per liter)
 */
export function carbonationToGramsPerLiter(
  carbonation: CarbonationType
): number {
  switch (carbonation.unit) {
    case "vols":
      return carbonation.value * 1.96;
    case "g/l":
      return carbonation.value;
    default:
      throw Error(`Unrecognized carbonation unit: ${carbonation.unit}`);
  }
}

// ============================================================================
// BITTERNESS CONVERSIONS
// ============================================================================

/**
 * Convert any BitternessType to IBUs
 */
export function bitternessToIBU(bitterness: BitternessType): number {
  switch (bitterness.unit) {
    case "IBUs":
      return bitterness.value;
    default:
      throw Error(`Unrecognized bitterness unit: ${bitterness.unit}`);
  }
}

// ============================================================================
// PERCENT CONVERSIONS
// ============================================================================

/**
 * Convert any PercentType to decimal (0-1 range)
 */
export function percentToDecimal(percent: PercentType): number {
  switch (percent.unit) {
    case "%":
      return percent.value / 100;
    default:
      throw Error(`Unrecognized percent unit: ${percent.unit}`);
  }
}

/**
 * Convert any PercentType to percentage value (0-100 range)
 */
export function percentToValue(percent: PercentType): number {
  switch (percent.unit) {
    case "%":
      return percent.value;
    default:
      throw Error(`Unrecognized percent unit: ${percent.unit}`);
  }
}

File: brewdio-core/src/units.rs
//! Unit conversion utilities for BeerJSON types.
//! All conversions support the complete set of units defined in the BeerJSON format.

use crate::beerjson_types::*;

// ============================================================================
// VOLUME CONVERSIONS
// ============================================================================

/// Convert any VolumeType to milliliters (ml)
pub fn volume_to_milliliters(volume: &VolumeType) -> f64 {
    match volume.unit {
        VolumeUnitType::Ml => volume.value,
        VolumeUnitType::L => volume.value * 1000.0,
        VolumeUnitType::Tsp => volume.value * 4.92892,
        VolumeUnitType::Tbsp => volume.value * 14.7868,
        VolumeUnitType::Floz => volume.value * 29.5735,
        VolumeUnitType::Cup => volume.value * 236.588,
        VolumeUnitType::Pt => volume.value * 473.176,
        VolumeUnitType::Qt => volume.value * 946.353,
        VolumeUnitType::Gal => volume.value * 3785.41,
        VolumeUnitType::Bbl => volume.value * 117347.77,
        VolumeUnitType::Ifloz => volume.value * 28.4131,
        VolumeUnitType::Ipt => volume.value * 568.261,
        VolumeUnitType::Iqt => volume.value * 1136.52,
        VolumeUnitType::Igal => volume.value * 4546.09,
        VolumeUnitType::Ibbl => volume.value * 163659.24,
    }
}

/// Convert any VolumeType to liters (l)
pub fn volume_to_liters(volume: &VolumeType) -> f64 {
    volume_to_milliliters(volume) / 1000.0
}

/// Convert any VolumeType to US gallons (gal)
pub fn volume_to_gallons(volume: &VolumeType) -> f64 {
    volume_to_milliliters(volume) / 3785.41
}

/// Convert any VolumeType to Imperial gallons (igal)
pub fn volume_to_imperial_gallons(volume: &VolumeType) -> f64 {
    volume_to_milliliters(volume) / 4546.09
}

// ============================================================================
// MASS CONVERSIONS
// ============================================================================

/// Convert any MassType to grams (g)
pub fn mass_to_grams(mass: &MassType) -> f64 {
    match mass.unit {
        MassUnitType::Mg => mass.value / 1000.0,
        MassUnitType::G => mass.value,
        MassUnitType::Kg => mass.value * 1000.0,
        MassUnitType::Lb => mass.value * 453.592,
        MassUnitType::Oz => mass.value * 28.3495,
    }
}

/// Convert any MassType to kilograms (kg)
pub fn mass_to_kilograms(mass: &MassType) -> f64 {
    mass_to_grams(mass) / 1000.0
}

/// Convert any MassType to pounds (lb)
pub fn mass_to_pounds(mass: &MassType) -> f64 {
    mass_to_grams(mass) / 453.592
}

/// Convert any MassType to ounces (oz)
pub fn mass_to_ounces(mass: &MassType) -> f64 {
    mass_to_grams(mass) / 28.3495
}

// ============================================================================
// TIME CONVERSIONS
// ============================================================================

/// Convert any TimeType to seconds (sec).
///
/// Returns f64 despite TimeType.value being i64, because all downstream
/// calculations (brewing formulas, unit conversions) operate in f64.
pub fn time_to_seconds(time: &TimeType) -> f64 {
    let v = time.value as f64;
    match time.unit {
        TimeUnitType::Sec => v,
        TimeUnitType::Min => v * 60.0,
        TimeUnitType::Hr => v * 3600.0,
        TimeUnitType::Day => v * 86400.0,
        TimeUnitType::Week => v * 604800.0,
    }
}

/// Convert any TimeType to minutes (min)
pub fn time_to_minutes(time: &TimeType) -> f64 {
    time_to_seconds(time) / 60.0
}

/// Convert any TimeType to hours (hr)
pub fn time_to_hours(time: &TimeType) -> f64 {
    time_to_seconds(time) / 3600.0
}

/// Convert any TimeType to days (day)
pub fn time_to_days(time: &TimeType) -> f64 {
    time_to_seconds(time) / 86400.0
}

// ============================================================================
// TEMPERATURE CONVERSIONS
// ============================================================================

/// Convert any TemperatureType to Celsius (C)
pub fn temperature_to_celsius(temp: &TemperatureType) -> f64 {
    match temp.unit {
        TemperatureUnitType::C => temp.value,
        TemperatureUnitType::F => (temp.value - 32.0) * (5.0 / 9.0),
    }
}

/// Convert any TemperatureType to Fahrenheit (F)
pub fn temperature_to_fahrenheit(temp: &TemperatureType) -> f64 {
    match temp.unit {
        TemperatureUnitType::C => temp.value * (9.0 / 5.0) + 32.0,
        TemperatureUnitType::F => temp.value,
    }
}

// ============================================================================
// COLOR CONVERSIONS
// ============================================================================

/// Convert SRM to EBC
pub fn srm_to_ebc(srm: f64) -> f64 {
    1.97 * srm
}

/// Convert EBC to SRM
pub fn ebc_to_srm(ebc: f64) -> f64 {
    ebc / 1.97
}

/// Convert Lovibond to SRM
pub fn lovibond_to_srm(lovibond: f64) -> f64 {
    1.3546 * lovibond - 0.76
}

/// Convert SRM to Lovibond
pub fn srm_to_lovibond(srm: f64) -> f64 {
    (srm + 0.76) / 1.3546
}

/// Convert any ColorType to SRM
pub fn color_to_srm(color: &ColorType) -> f64 {
    match color.unit {
        ColorUnitType::Srm => color.value,
        ColorUnitType::Ebc => ebc_to_srm(color.value),
        ColorUnitType::Lovi => lovibond_to_srm(color.value),
    }
}

/// Convert any ColorType to EBC
pub fn color_to_ebc(color: &ColorType) -> f64 {
    srm_to_ebc(color_to_srm(color))
}

/// Convert any ColorType to Lovibond
pub fn color_to_lovibond(color: &ColorType) -> f64 {
    srm_to_lovibond(color_to_srm(color))
}

// ============================================================================
// PRESSURE CONVERSIONS
// ============================================================================

/// Convert any PressureType to kilopascals (kPa)
pub fn pressure_to_kilopascals(pressure: &PressureType) -> f64 {
    match pressure.unit {
        PressureUnitType::KPa => pressure.value,
        PressureUnitType::Psi => pressure.value * 6.89476,
        PressureUnitType::Bar => pressure.value * 100.0,
    }
}

/// Convert any PressureType to PSI
pub fn pressure_to_psi(pressure: &PressureType) -> f64 {
    pressure_to_kilopascals(pressure) / 6.89476
}

/// Convert any PressureType to bar
pub fn pressure_to_bar(pressure: &PressureType) -> f64 {
    pressure_to_kilopascals(pressure) / 100.0
}

// ============================================================================
// GRAVITY CONVERSIONS
// ============================================================================

/// Convert any GravityType to specific gravity (sg)
pub fn gravity_to_sg(gravity: &GravityType) -> f64 {
    match gravity.unit {
        GravityUnitType::Sg => gravity.value,
        // Plato to SG: SG = 1 + (Plato / (258.6 - ((Plato / 258.2) * 227.1)))
        GravityUnitType::Plato => {
            1.0 + gravity.value / (258.6 - (gravity.value / 258.2) * 227.1)
        }
        // Brix to SG (using same formula as Plato, as they're very similar)
        GravityUnitType::Brix => {
            1.0 + gravity.value / (258.6 - (gravity.value / 258.2) * 227.1)
        }
    }
}

/// Convert any GravityType to Plato
pub fn gravity_to_plato(gravity: &GravityType) -> f64 {
    match gravity.unit {
        GravityUnitType::Sg => {
            // SG to Plato: Plato = -616.868 + 1111.14 * SG - 630.272 * SG^2 + 135.997 * SG^3
            let sg = gravity.value;
            -616.868 + 1111.14 * sg - 630.272 * sg.powi(2) + 135.997 * sg.powi(3)
        }
        GravityUnitType::Plato => gravity.value,
        GravityUnitType::Brix => gravity.value,
    }
}

/// Convert any GravityType to Brix
pub fn gravity_to_brix(gravity: &GravityType) -> f64 {
    gravity_to_plato(gravity)
}

// ============================================================================
// CONCENTRATION CONVERSIONS
// ============================================================================

/// Convert any ConcentrationType to mg/l (milligrams per liter)
pub fn concentration_to_mg_per_l(concentration: &ConcentrationType) -> f64 {
    match concentration.unit {
        ConcentrationUnitType::MgL => concentration.value,
        ConcentrationUnitType::Ppm => concentration.value, // ppm ≡ mg/l for aqueous solutions
        ConcentrationUnitType::Ppb => concentration.value / 1000.0,
    }
}

/// Convert any ConcentrationType to ppm (parts per million)
pub fn concentration_to_ppm(concentration: &ConcentrationType) -> f64 {
    concentration_to_mg_per_l(concentration)
}

/// Convert any ConcentrationType to ppb (parts per billion)
pub fn concentration_to_ppb(concentration: &ConcentrationType) -> f64 {
    concentration_to_mg_per_l(concentration) * 1000.0
}

// ============================================================================
// SPECIFIC VOLUME CONVERSIONS
// ============================================================================

/// Convert any SpecificVolumeType to l/kg (liters per kilogram)
pub fn specific_volume_to_l_per_kg(specific_volume: &SpecificVolumeType) -> f64 {
    match specific_volume.unit {
        SpecificVolumeUnitType::LKg => specific_volume.value,
        SpecificVolumeUnitType::LG => specific_volume.value * 1000.0,
        SpecificVolumeUnitType::QtLb => specific_volume.value * 2.08635,
        SpecificVolumeUnitType::GalLb => specific_volume.value * 8.34540,
        SpecificVolumeUnitType::GalOz => specific_volume.value * 133.526,
        SpecificVolumeUnitType::FlozOz => specific_volume.value * 1.04318,
        SpecificVolumeUnitType::M3Kg => specific_volume.value * 1000.0,
        SpecificVolumeUnitType::Ft3Lb => specific_volume.value * 62.4279,
    }
}

/// Convert any SpecificVolumeType to qt/lb (quarts per pound)
pub fn specific_volume_to_qt_per_lb(specific_volume: &SpecificVolumeType) -> f64 {
    specific_volume_to_l_per_kg(specific_volume) / 2.08635
}

/// Convert any SpecificVolumeType to gal/lb (gallons per pound)
pub fn specific_volume_to_gal_per_lb(specific_volume: &SpecificVolumeType) -> f64 {
    specific_volume_to_l_per_kg(specific_volume) / 8.34540
}

/// Convert any SpecificVolumeType to gal/kg (gallons per kilogram)
pub fn specific_volume_to_gal_per_kg(specific_volume: &SpecificVolumeType) -> f64 {
    specific_volume_to_l_per_kg(specific_volume) * 0.264172
}

// ============================================================================
// ACIDITY CONVERSIONS
// ============================================================================

/// Convert any AcidityType to pH value
pub fn acidity_to_ph(acidity: &AcidityType) -> f64 {
    match acidity.unit {
        AcidityUnitType::PH => acidity.value,
    }
}

// ============================================================================
// CARBONATION CONVERSIONS
// ============================================================================

/// Convert any CarbonationType to volumes of CO2
pub fn carbonation_to_volumes(carbonation: &CarbonationType) -> f64 {
    match carbonation.unit {
        CarbonationUnitType::Vols => carbonation.value,
        // g/l to volumes: 1 volume = 1.96 g/l at STP
        CarbonationUnitType::GL => carbonation.value / 1.96,
    }
}

/// Convert any CarbonationType to g/l (grams per liter)
pub fn carbonation_to_grams_per_liter(carbonation: &CarbonationType) -> f64 {
    match carbonation.unit {
        CarbonationUnitType::Vols => carbonation.value * 1.96,
        CarbonationUnitType::GL => carbonation.value,
    }
}

// ============================================================================
// BITTERNESS CONVERSIONS
// ============================================================================

/// Convert any BitternessType to IBUs
pub fn bitterness_to_ibu(bitterness: &BitternessType) -> f64 {
    match bitterness.unit {
        BitternessUnitType::IbUs => bitterness.value,
    }
}

// ============================================================================
// PERCENT CONVERSIONS
// ============================================================================

/// Convert any PercentType to decimal (0-1 range)
pub fn percent_to_decimal(percent: &PercentType) -> f64 {
    match percent.unit {
        PercentUnitType::X => percent.value / 100.0,
    }
}

/// Convert any PercentType to percentage value (0-100 range)
pub fn percent_to_value(percent: &PercentType) -> f64 {
    match percent.unit {
        PercentUnitType::X => percent.value,
    }
}
